docs/extensibility/extend-rest-api-add-data.md

```diff
@@ -0,0 +1,289 @@
+ 'type' => 'string'
+ 'readonly' => true,
+ ]
+ ]
+}
+```
+
+Data callback and Schema callback can also receive parameters:
+
+```PHP
+
+function my_cart_item_callback( $cart_item ) {
+ $product = $cart_item['data'];
+ if ( is_my_custom_product_type( $product ) ) {
+ $custom_value = get_custom_value( $product );
+ return [
+ 'custom-key' => $custom_value;
+ ]
+ }
+}
+
+```
+
+## API Definition
+
+- `ExtendRestAPI::register_endpoint_data`: Used to register data to a custom endpoint. It takes an array of arguments:
+
+| Attribute | Type | Required | Description |
+| :-------- | :----- | :------: | :------------------------------------ |
+| `endpoint` | string | Yes | The endpoint you're trying to extend. It is suggested that you use the `::IDENTIFIER` available on the route Schema class to avoid typos. |
+| `namespace` | string | Yes | Your plugin namespace, the data will be available under this namespace in the StoreAPI response. |
+| `data_callback` | callback | Yes | A callback that returns an array with your data. |
+| `schema_callback` | callback | Yes | A callback that returns the shape of your data. |
+| `data_type` | string | No (default: `ARRAY_A` ) | The type of your data. If you're adding an object (key => values), it should be `ARRAY_A`. If you're adding a list of items, it should be `ARRAY_N`. |
+
+## Putting it all together
+
+This is a complete example that shows how you can register contextual WooCommerce Subscriptions data in each cart item (simplified).
+
+This example uses Formatters, another property documented here (TBD).
+
+```php
+<?php
+/**
+ * WooCommerce Subscriptions Extend Store API.
+ *
+ * A class to extend the store public API with subscription related data
+ * for each subscription item
+ *
+ * @package WooCommerce Subscriptions
+ */
+
+use Automattic\WooCommerce\Blocks\Package;
+use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
+use Automattic\WooCommerce\Blocks\StoreApi\SchemaController;
+use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartItemSchema;
+
+if ( class_exists( 'Package' ) && version_compare( Package::get_version(), '4.8.0', '>=' ) ) {
+	// This class needs to run after WooCommerce Blocks is ready.
+	add_action( 'woocommerce_blocks_loaded', array( 'WC_Subscriptions_Extend_Store_Endpoint', 'init' ) );
+}
+
+class WC_Subscriptions_Extend_Store_Endpoint {
+
+	/**
+	 * Stores Rest Schema Controller.
+	 *
+	 * @var ExtendRestApi
+	 */
+	private static $schema;
+
+	/**
+	 * Stores Rest Extending instance.
+	 *
+	 * @var ExtendRestApi
+	 */
+	private static $extend;
+
+	/**
+	 * Plugin Identifier, unique to each plugin.
+	 *
+	 * @var string
+	 */
+	const IDENTIFIER = 'subscriptions';
+
+	/**
+	 * Bootstraps the class and hooks required data.
+	 *
+	 * @since WCBLOCKS-DEV
+	 */
+	public static function init() {
+		self::$schema = Package::container()->get( SchemaController::class );
+		self::$extend = Package::container()->get( ExtendRestApi::class );
+		self::extend_store();
+	}
```

Comment:
Nit, to promote DI patterns a bit more, it'd be good to pass these in on the `init` method rather than instantiating in the class:

```suggestion
	public static function init( SchemaController $schema_controller, ExtendRestAPI $extend_rest_api ) {
		self::$schema = $schema_controller;
		self::$extend = $extend_rest_api;
		self::extend_store();
	}
```

There's at least few reasons for this:

- Direct container access should typically happen in as few places as possible (typically in bootstrapping) because usage elsewhere is typically bad usage of the service locator pattern (which we have here).
- Unit Testing this would be much more difficult. While it'd be possible to swap `SchemaController` and `ExtendRestAPI` mocks into the container, it's not trivial and requires understanding how the container works. Instead, with passing them in as dependencies, they could be mocked much more easily for testing.
- The class is tightly coupled to the availability of `Package::container()`. While your example does make sure it's available _outside this class_ by only initializing it on the `woocommerce_blocks_loaded` action, note _that happens outside the class_. Let's make that more explicit and also pass in the SchemaController and ExtendRestAPI _outside the class_ in the same logic also testing for the availability.